### PR TITLE
Added webinterface option for C++ chat client, small Rust config changes

### PR DESCRIPTION
**Pull request by kieranjazy**
Added C++ web interface available by default. Can be disabled by removing --webinterface from Enarx.toml. Uses port 8080. Made small changes to Rust files to fix broken file locations

### CODE REVIEW

**Comment by rjzak** [comment]
A better way to do things might have been to rebase https://github.com/enarx/codex/pull/69 and update.

* So you make your changes.
* `git commit -as` Commit you new changes
* `git rebase -i HEAD~2` Compare last two commits
* Replace the `pick` with `f`
* Save, and git will "fix-up" the commit with the next commit, keeping the history clean.
* `git push -f origin <my branch>` to force push, and if you have an open PR, it'll be automatically updated.

Also, you should probably work in a new branch instead of `main`. This is easy:
* `git checkout main`
* `git pull origin main` to get any changes which might be in your repo
* `git remote add upstream https://github.com/enarx/codex.git` add the main repo as the "upstream"
* `git fetch upstream` get the upstream history
* `git rebase upstream/main` get the latest changes from upstream, add to your `main`
* `git push origin main` freshen up your copy of main
* Make some changes to your code, don't commit yet!
* `git checkout -b <my_new_branch>`
* `git push origin <my_new_branch>`
* Open PR against your branch
* Updates (shown above) will automatically update your PR, don't need to close and make new ones!

If you have any questions, let me know, or ask in the [Enarx chat](https://chat.enarx.dev)

**Comment by kieranjazy** [comment]
Would it be OK to merge this PR as it is now or should I add it all to a new branch?

**Comment by rjzak** [comment]
Don't worry about having a new branch this time. Hopefully @rvolosatovs will be able to take a look.

**Comment by kieranjazy** [comment]
I've it fixed on local, was using your python script to generate the webpage assets but didn't realise it'd break something so I've included it. I'll upload final commit now and solve merge issues

**Comment by rjzak** [comment]
That's fine for now, but the Nix code should be updated to generate `assets.h` instead of including it.

**Comment by kieranjazy** [comment]
> That's fine for now, but the Nix code should be updated to generate `assets.h` instead of including it.

Ah ok, I'll look into Nix, haven't used it before

**Comment by nickvidal** [comment]
Hi @kieranjazy, I tested your PR and it works great. Thanks!

**Comment by kieranjazy** [comment]
Thanks Nick, just struggling to get Nix to run a Python script for me, still working on it

**Comment by rvolosatovs** [comment]
> That's fine for now, but the Nix code should be updated to generate `assets.h` instead of including it.

I don't think we need that, at least it should not be a blocker for this PR, this warrants more discussion. Can we file a "goldplating" tech debt issue about it and get this merged?

File: demos/chat-client/Enarx.toml [hunk 1]
```diff
@@ -13,3 +17,10 @@ kind = "connect"
 port = 50000
 prot = "tcp"
 host = "127.0.0.1"
+
+[[files]]
+name = "web"
+kind = "listen"
+port = 8080
+prot = "tcp"
```
Comment (rvolosatovs) [inline demos/chat-client/Enarx.toml:25]: let's use tls here

File: demos/chat-client/Enarx.toml [hunk 2]
```diff
@@ -13,3 +17,10 @@ kind = "connect"
 port = 50000
 prot = "tcp"
 host = "127.0.0.1"
+
+[[files]]
+name = "web"
+kind = "listen"
+port = 8080
```
Comment (rvolosatovs) [inline demos/chat-client/Enarx.toml:24]: Let's use something more unique.
`50000` is already defined as the peer endpoint
How about `50010` for the web interface?

File: demos/chat-client/c++/README.md [hunk 3]
```diff
@@ -1 +1,2 @@
 Compile to WASM first, then run with --wasmcfgfile directed to Enarx.toml
+Use the --webinterface argument in the Enarx.toml to switch to a web interface, at localhost:8080
```
Comment (rvolosatovs) [inline demos/chat-client/c++/README.md:2]: Is there a reason not to enable it by default?

File: demos/chat-client/c++/compile_run.sh [hunk 4]
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash
+
+# WASI_SDK_PATH defined in https://github.com/WebAssembly/wasi-sdk
+$WASI_SDK_PATH/bin/clang++ main.cpp --sysroot $WASI_SDK_PATH/share/wasi-sysroot -o main.wasm &&
```
Comment (rvolosatovs) [inline demos/chat-client/c++/compile_run.sh:4]: I don't think this belongs in the repository, looks like a local development environment tooling

File: demos/chat-client/c++/main.cpp [hunk 5]
```diff
@@ -3,24 +3,129 @@
...
+// Pass --webinterface in Enarx.toml to use webinterface, uses std::in otherwise
+
+int main(int argc, char *argv[])
+{
+    if (argc > 2) {
+        std::cout << "Too many arguments provided" << std::endl;
+        return 1;
+    }
+
+    bool isWebInterface = false;
```
Comment (rvolosatovs) [inline demos/chat-client/c++/main.cpp:33]: I think this should just be enabled by default, we can add functionality to disable it later if we need it

File: demos/chat-client/c++/process_assets.py [hunk 6]
```diff
@@ -0,0 +1,18 @@
+#!/usr/bin/env nix-shell
+#!nix-shell -i python3 -p "python3.withPackages([])"
+
+# Python script from https://github.com/rjzak/web_wordpuzzle
+
+ASSETS = ( ("jquery", open("assets/jquery-3.6.0.min.js", "rb").read()), ("chat", open("assets/chat.html", "rb").read()), )
```
Comment (rvolosatovs) [inline demos/chat-client/c++/process_assets.py:6]: Why add another development dependency for a simple script like this? Can't we use C++ functionality to take care of this directly at compilation time?
Could this be just a simple C++ program, which is run at build time?

File: demos/chat-client/c++/process_assets.py [hunk 7]
```diff
@@ -0,0 +1,18 @@
+#!/usr/bin/env nix-shell
```
Comment (rvolosatovs) [inline demos/chat-client/c++/process_assets.py:1]: We should not make `nix-shell` a dependency here
```suggestion
#!/usr/bin/env python
```

**Review by rvolosatovs — CHANGES_REQUESTED** [review]

File: demos/chat-client/c++/README.md [hunk 8]
```diff
@@ -1 +1,6 @@
 Compile to WASM first, then run with --wasmcfgfile directed to Enarx.toml
+<<<<<<< HEAD
```
Comment (rvolosatovs) [inline demos/chat-client/c++/README.md:2]: Looks like an artifact from an unfinished rebase

File: demos/chat-client/c++/assets.h [hunk 9]
```diff
@@ -0,0 +1,3 @@
...x65,0x68,0x6F,0x6C,0x64,0x65,0x72,0x3D,0x27,0x54,0x79,0x70,0x65,0x20,0x61,0x20,0x6D,0x65,0x73,0x73,0x61,0x67,0x65,0x20,0x2E,0x2E,0x2E,0x27,0x20,0x61,0x75,0x74,0x6F,0x63,0x6F,0x6D,0x70,0x6C,0x65,0x74,0x65,0x3D,0x27,0x6F,0x66,0x66,0x27,0x3E,0x0A,0x20,0x20,0x20,0x20,0x20,0x20,0x20,0x20,0x20,0x20,0x20,0x20,0x20,0x20,0x20,0x20,0x3C,0x62,0x75,0x74,0x74,0x6F,0x6E,0x20,0x63,0x6C,0x61,0x73,0x73,0x3D,0x27,0x63,0x68,0x61,0x74,0x2D,0x73,0x75,0x62,0x6D,0x69,0x74,0x27,0x3E,0x3C,0x69,0x20,0x63,0x6C,0x61,0x73,0x73,0x3D,0x27,0x6D,0x61,0x74,0x65,0x72,0x69,0x61,0x6C,0x2D,0x69,0x63,0x6F,0x6E,0x73,0x27,0x3E,0x73,0x65,0x6E,0x64,0x3C,0x2F,0x69,0x3E,0x3C,0x2F,0x62,0x75,0x74,0x74,0x6F,0x6E,0x3E,0x0A,0x20,0x20,0x20,0x20,0x20,0x20,0x20,0x20,0x20,0x20,0x20,0x20,0x3C,0x2F,0x64,0x69,0x76,0x3E,0x0A,0x0A,0x20,0x20,0x20,0x20,0x20,0x20,0x20,0x20,0x3C,0x2F,0x64,0x69,0x76,0x3E,0x0A,0x20,0x20,0x20,0x20,0x3C,0x2F,0x64,0x69,0x76,0x3E,0x0A,0x20,0x20,0x20,0x20,0x3C,0x73,0x63,0x72,0x69,0x70,0x74,0x20,0x73,0x72,0x63,0x3D,0x22,0x2E,0x2F,0x73,0x63,0x72,0x69,0x70,0x74,0x2E,0x6A,0x73,0x22,0x3E,0x3C,0x2F,0x73,0x63,0x72,0x69,0x70,0x74,0x3E,0x0A,0x0A,0x3C,0x2F,0x62,0x6F,0x64,0x79,0x3E,0x0A,0x0A,0x3C,0x2F,0x68,0x74,0x6D,0x6C,0x3E};
```
Comment (rvolosatovs) [inline demos/chat-client/c++/assets.h:1]: There should be a comment at the top of this file saying that this is auto-generated and pointing to the source.
Could you add the script to the repo as well? We don't have to make it a part of the build pipeline, but there should be a way to reproduce this

File: flake.nix [hunk 10]
```diff
@@ -108,7 +108,7 @@
         }
         ''
           mkdir -p "$out/bin"
-          $CXX -Wall -pedantic ${demos.chat-client.cpp.src} \
+          $CXX -Wall -I"./demos/chat-client/c++/assets.h" -pedantic ${demos.chat-client.cpp.src} \
```
Comment (rvolosatovs) [inline flake.nix:111]:
```suggestion
          $CXX -Wall -I"${demos.chat-client.cpp.src}/assets.h" -pedantic ${demos.chat-client.cpp.src}/fibonacci.cpp \
```

Will allow `nix` to access `assets.h` if you set `demos.fibonacci.cpp.src = ./demos/fibonacci/c++;` on L37

File: demos/chat-client/Enarx.toml (hunk 1 above)
Comment (rvolosatovs) [reply demos/chat-client/Enarx.toml:21]:
```suggestion
prot = "tls"
```

File: demos/chat-client/c++/assets/script.js [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+    chatBtn = document.querySelector(".chat-btn")
+    chatSubmit = document.querySelector(".chat-submit")
+    chatHeader = document.querySelector(".chat-header")
+    chatArea = document.querySelector(".chat-area")
+    chatInput = document.querySelector(".chat-input")
+    stopButton = document.querySelector(".stop-btn")
+    expandWindow = document.querySelector(".expand-chat-window")
+    root = document.documentElement;
+    chatPopup.style.display = "none"
+    var host = "http://localhost:50010"
```
Comment (rvolosatovs) [inline demos/chat-client/c++/assets/script.js:15]:
```suggestion
    var host = "https://localhost:50010"
```

File: demos/chat-client/c++/assets/script.js [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+    }
+}
+
+function detectMob() {
+    return ((window.innerHeight <= 800) && (window.innerWidth <= 600));
+}
+
+function createChatBot() {
+
+    host = "http://localhost:50010";
```
Comment (rvolosatovs) [inline demos/chat-client/c++/assets/script.js:123]: unused?

File: demos/chat-client/c++/assets/script.js [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+            // chatPopup.style.transition = "none"
+        expandWindow.innerHTML = `<i class='material-icons'>close</i>`
+    }
+}
+
+function detectMob() {
+    return ((window.innerHeight <= 800) && (window.innerWidth <= 600));
+}
+
+function createChatBot() {
```
Comment (rvolosatovs) [inline demos/chat-client/c++/assets/script.js:121]: There should not be a bot

**Comment by rvolosatovs** [comment]
@kieranjazy I believe the outdated `wasi-libc` issue is now solved, but the code still fails to compile with:
```
chat-client-wasm32-unknown-wasi> /nix/store/fxasr4jvncasd20wkisj6wf78gyww84y-c++/main.cpp:17:21: warning: unused variable 'HTTP_NOT_FOUND' [-Wunused-const-variable]
chat-client-wasm32-unknown-wasi> const unsigned char HTTP_NOT_FOUND[14] = "HTTP/1.0 404\n";
chat-client-wasm32-unknown-wasi>                     ^
chat-client-wasm32-unknown-wasi> /nix/store/fxasr4jvncasd20wkisj6wf78gyww84y-c++/main.cpp:19:21: warning: unused variable 'CONTENT_TYPE_PLAIN' [-Wunused-const-variable]
chat-client-wasm32-unknown-wasi> const unsigned char CONTENT_TYPE_PLAIN[27] = "Content-Type: text/plain\n\n";
chat-client-wasm32-unknown-wasi>                     ^
chat-client-wasm32-unknown-wasi> 2 warnings generated.
chat-client-wasm32-unknown-wasi> /build/.attr-0l2nkwhif96f51f4amnlf414lhl4rv9vh8iffyp431v6s28gsr90: line 3: -o: command not found
```

File: flake.nix [hunk 14]
```diff
@@ -128,7 +129,7 @@
         }
         ''
           mkdir -p "$out/bin"
-          $CXX -Wall -pedantic ${demos.chat-client.cpp.src} \
+          $CXX -Wall -I"${demos.chat-client.cpp.src}/assets.h" -pedantic ${demos.chat-client.cpp.src}/main.cpp
```
Comment (rvolosatovs) [inline flake.nix:132]:
```suggestion
          $CXX -Wall -I"${demos.chat-client.cpp.src}/assets.h" -pedantic ${demos.chat-client.cpp.src}/main.cpp \
```

File: demos/chat-server/rust/.cargo/config [hunk 15]
```diff
@@ -2,5 +2,5 @@
 target = "wasm32-wasi"
 
 [target.wasm32-wasi]
-runner = ["enarx", "run", "--wasmcfgfile", "Enarx.toml"]
+runner = ["enarx", "run", "--wasmcfgfile", "../Enarx.toml"]
```
Comment (rvolosatovs) [inline demos/chat-server/rust/.cargo/config:5]: Can we avoid this diff?
Is it possible for you to extract this into a separate PR?

File: demos/chat-server/rust/Cargo.lock [hunk 16]
```diff
@@ -42,6 +42,17 @@ version = "1.0.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "baf1de4339761588bc0619e3cbc0120ee582ebb74b53b4efbf79117bd2da40fd"
 
+[[package]]
```
Comment (rvolosatovs) [inline demos/chat-server/rust/Cargo.lock:45]: Same as above

**Review by rvolosatovs — CHANGES_REQUESTED** [review]

**Review by rvolosatovs — APPROVED** [review]

**Pull request merged**